pkg/controller/statefulset/metrics.go

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package statefulset
+
+import (
+	"sync"
+
+	apps "k8s.io/api/apps/v1"
+	"k8s.io/component-base/metrics"
+)
+
+// StatefulSet is the subsystem name used by this package.
+const StatefulSet = "statefulset"
+
+var (
+	deletedPolicyCounter = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Subsystem:      StatefulSet,
+			Name:           "statefulset_non_retain_when_deleted_policy",
```

Comment:
At this point the only policy that is non-retain is delete, so the question is a little academic.

My motivation here is that "retain" is the current default behavior. The point of this metric as articulated in the KEP is to provide a way for cluster operators to quickly see if the new feature is being used.

Maybe it would be better to have this be something like "statefulset_reconcile_count", and then have fields like delete_policy_retain, delete_policy_delete, etc? More information using fields, but also more setup and configuration to figure out what is interesting rather than a single metric.

I can see arguments either way, I leaned towards simpler metrics.

Reply:
It's a bit of an anti-pattern to embed what can be a label value into the name of a metric. In this case, `when deleted` is a state, as is `when scaled`. So I would instead consolidate those metrics into a single one with a `state` label with possible values of `deleted` and `scaled`.

Reply:
done, thx for the feedback. LMK if that's not what you had in mind.

Reply:
yes this is much better . Although I am wondering these are still things which are present in pod resource object . If I really cared about these metrics , I can easily run https://github.com/kubernetes/kube-state-metrics to get prometheus version of this state and any future state i care about which is present in pod/statefulset yaml.

I remember seeing prod readiness doc asking questions about this. I wonder if we can settle on for asking this question by
- i know this feature is enabled if kubectl returns xyz
- i know this feature is enabled if kube-state-metrics returns blah

Thoughts @kow3ns  @soltysh  @johnbelamaric (for prod readiness)

Reply:
kube-state-metrics is a great solution, I agree it's not really scalable to be adding little metrics like these for each feature.